core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.calcite.rel.type.RelDataTypeSystem;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.Table;
+
+import org.junit.Test;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Arrays;
+import java.util.Collections;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Test for querying Java Object array table.
+ */
+public class ObjectArrayTableTest {
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-1703">[CALCITE-1703]
+   * Functions on TIME and TIMESTAMP column can throw ClassCastException</a>. */
+  @Test public void testJavaSqlDateColumnReference() throws SQLException {
+
+    try (
+        Connection connection = DriverManager.getConnection("jdbc:calcite:");
+        CalciteConnection calciteConnection = connection.unwrap(CalciteConnection.class)
+    ) {
+      final SchemaPlus rootSchema = calciteConnection.getRootSchema();
+
+      final JavaTypeFactoryImpl typeFactory =
+          new JavaTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+
+      final RelDataType rowType = typeFactory.createStructType(
+          Arrays.asList(
+              typeFactory.createJavaType(java.sql.Date.class),
+              typeFactory.createJavaType(java.sql.Time.class),
+              typeFactory.createJavaType(java.sql.Timestamp.class)
+          ),
+          Arrays.asList("dt", "tm", "ts")
+      );
+
+      final Enumerable<Object[]> enumerable =
+          Linq4j.asEnumerable(
+              Collections.singletonList(new Object[]{
+                  java.sql.Date.valueOf("2018-12-14"),
+                  java.sql.Time.valueOf("18:29:34"),
+                  java.sql.Timestamp.valueOf("2018-12-14 18:29:34.123")}));
+
+      final Table table = new AbstractQueryableTable(Object[].class) {
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        @Override public Queryable<Object[]> asQueryable(QueryProvider queryProvider,
+                                               SchemaPlus schema, String tableName) {
+          return EnumerableDefaults.asOrderedQueryable(enumerable);
+        }
+
+        @Override public RelDataType getRowType(RelDataTypeFactory typeFactory) {
+          return rowType;
+        }
+      };
+
+      rootSchema.add("java_sql_date_types", table);
+
+      // 'extract' function expects numeric date representation.
+      // RexToLixTranslator.convert() handles Java type to numeric conversion.
+      final Statement statement = connection.createStatement();
+      String sql = "select extract(year from \"dt\"),"
+          + " extract(hour from \"tm\"),"
+          + " extract(day from \"ts\") from \"java_sql_date_types\"";
+
+      ResultSet resultSet = statement.executeQuery(sql);
+      assertTrue(resultSet.next());
+      assertEquals(2018, resultSet.getInt(1));
```

Comment:
Would you please add a message to `assertEquals`?
Assert failure should look like a decent bug report, and current `assertEquals(2018, resultSet.getInt(1));` would produce something like `expected 2018 got 0`.

The other way to look at "message" parameter: would you be able to tell the nature of the bug by looking at error message+stacktrace only?

Reply:
Makes sense. I've added messages. Thanks!